File: android/app/src/main/java/com/soyxan/sidesubs/SubtitleParser.java
package com.soyxan.sidesubs;

import java.util.ArrayList;
import java.util.List;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

import com.soyxan.sidesubs.PlexModels.Cue;

final class SubtitleParser {
    private static final Pattern SRT_TIME = Pattern.compile(
        "(\\d{1,2}):(\\d{2}):(\\d{2})[,.](\\d{1,3})\\s*-->\\s*" +
        "(\\d{1,2}):(\\d{2}):(\\d{2})[,.](\\d{1,3})"
    );
    private static final Pattern ASS_TAG = Pattern.compile("\\{[^}]*}");

    private SubtitleParser() {}

    static List<Cue> parse(String text) {
        if (text == null || text.isEmpty()) return new ArrayList<>();
        List<Cue> ass = parseAss(text);
        if (!ass.isEmpty()) return ass;
        return parseSrt(text);
    }

    private static List<Cue> parseAss(String text) {
        List<Cue> cues = new ArrayList<>();
        String[] lines = text.replace("\r", "").split("\n");
        for (String line : lines) {
            if (!line.startsWith("Dialogue:")) continue;
            String body = line.substring("Dialogue:".length()).trim();
            String[] fields = body.split(",", 10);
            if (fields.length < 10) continue;
            Double start = parseAssTime(fields[1]);
            Double end = parseAssTime(fields[2]);
            if (start == null || end == null || end < start) continue;
            String value = cleanAssText(fields[9]);
            if (!value.isEmpty()) cues.add(new Cue(start, end, value));
        }
        return cues;
    }

    private static String cleanAssText(String value) {
        String cleaned = value
            .replace("\\N", "\n")
            .replace("\\n", "\n")
            .replace("\\h", " ");
        cleaned = ASS_TAG.matcher(cleaned).replaceAll("");
        return cleaned.trim();
    }

    private static Double parseAssTime(String value) {
        try {
            String[] p = value.trim().split(":");
            if (p.length != 3) return null;
            return Integer.parseInt(p[0]) * 3600.0
                + Integer.parseInt(p[1]) * 60.0
                + Double.parseDouble(p[2].replace(',', '.'));
        } catch (Exception ignored) {
            return null;
        }
    }

    private static List<Cue> parseSrt(String text) {
        List<Cue> cues = new ArrayList<>();
        String[] lines = text.replace("\r", "").split("\n");
        for (int i = 0; i < lines.length; i++) {
            Matcher matcher = SRT_TIME.matcher(lines[i].trim());
            if (!matcher.matches()) continue;

            double start = srtSeconds(matcher, 1);
            double end = srtSeconds(matcher, 5);
            StringBuilder value = new StringBuilder();
            int j = i + 1;
            while (j < lines.length && !lines[j].trim().isEmpty()) {
                if (value.length() > 0) value.append('\n');
                value.append(lines[j].trim());
                j++;
            }
            String cleaned = value.toString()
                .replaceAll("<[^>]+>", "")
                .trim();
            if (!cleaned.isEmpty() && end >= start) {
                cues.add(new Cue(start, end, cleaned));
            }
            i = j;
        }
        return cues;
    }

    private static double srtSeconds(Matcher m, int offset) {
        int h = Integer.parseInt(m.group(offset));
        int min = Integer.parseInt(m.group(offset + 1));
        int sec = Integer.parseInt(m.group(offset + 2));
        String msText = m.group(offset + 3);
        int ms = Integer.parseInt(msText);
        if (msText.length() == 1) ms *= 100;
        else if (msText.length() == 2) ms *= 10;
        return h * 3600.0 + min * 60.0 + sec + ms / 1000.0;
    }
}

File: android/app/src/main/java/com/soyxan/sidesubs/SubtitleParser.kt
package com.soyxan.sidesubs

private val SRT_TIME = Regex(
    """(\d{1,2}):(\d{2}):(\d{2})[,.](\d{1,3})\s*-->\s*(\d{1,2}):(\d{2}):(\d{2})[,.](\d{1,3})"""
)
private val ASS_TAG = Regex("""\{[^}]*\}""")
private val HTML_TAG = Regex("""<[^>]+>""")

object SubtitleParser {
    fun parse(text: String?): List<Cue> {
        if (text.isNullOrEmpty()) return emptyList()
        val ass = parseAss(text)
        return if (ass.isNotEmpty()) ass else parseSrt(text)
    }

    private fun parseAss(text: String): List<Cue> = buildList {
        text.replace("\r", "").lineSequence().forEach { line ->
            if (!line.startsWith("Dialogue:")) return@forEach
            val fields = line.removePrefix("Dialogue:").trim().split(",", limit = 10)
            if (fields.size < 10) return@forEach

            val start = parseAssTime(fields[1]) ?: return@forEach
            val end = parseAssTime(fields[2]) ?: return@forEach
            if (end < start) return@forEach

            val value = fields[9]
                .replace("\\N", "\n")
                .replace("\\n", "\n")
                .replace("\\h", " ")
                .replace(ASS_TAG, "")
                .trim()

            if (value.isNotEmpty()) add(Cue(start, end, value))
        }
    }

    private fun parseAssTime(value: String): Double? = runCatching {
        val parts = value.trim().split(":")
        require(parts.size == 3)
        parts[0].toInt() * 3600.0 +
            parts[1].toInt() * 60.0 +
            parts[2].replace(',', '.').toDouble()
    }.getOrNull()

    private fun parseSrt(text: String): List<Cue> {
        val lines = text.replace("\r", "").split("\n")
        val cues = mutableListOf<Cue>()
        var i = 0

        while (i < lines.size) {
            val match = SRT_TIME.matchEntire(lines[i].trim())
            if (match == null) {
                i++
                continue
            }

            val start = srtSeconds(match, 1)
            val end = srtSeconds(match, 5)
            val body = mutableListOf<String>()
            var j = i + 1
            while (j < lines.size && lines[j].isNotBlank()) {
                body += lines[j].trim()
                j++
            }
            val value = body.joinToString("\n").replace(HTML_TAG, "").trim()
            if (value.isNotEmpty() && end >= start) cues += Cue(start, end, value)
            i = j + 1
        }
        return cues
    }

    private fun srtSeconds(match: MatchResult, offset: Int): Double {
        val g = match.groupValues
        val hours = g[offset].toInt()
        val minutes = g[offset + 1].toInt()
        val seconds = g[offset + 2].toInt()
        val fraction = g[offset + 3]
        val millis = when (fraction.length) {
            1 -> fraction.toInt() * 100
            2 -> fraction.toInt() * 10
            else -> fraction.toInt()
        }
        return hours * 3600.0 + minutes * 60.0 + seconds + millis / 1000.0
    }
}
